### PR TITLE
build: disable Seastar exception hack

### PR DESCRIPTION
**Pull request by avikivity**
In [1], Seastar started to bypass a lock in libgcc's exception throwing mechanism to allow scalability on large machines. The problem is documented in [2] and reported as fixed.

In [3], testing results on a 2s96c192t machine are reported. The problem appears indeed fixed with gcc 14's runtime (which we use, even though we build with clang).

Given the new results, we can safely drop the exception scalability hack. As [1] states that the hack causes the loss of a translation cache, we may gain some performance this way.

With that, we disable the cache by defining some random macros.

[1] https://github.com/scylladb/seastar/commit/464f5e3ae43b366b05573018fc46321863bf2fae
[2] https://gcc.gnu.org/bugzilla/show_bug.cgi?id=71744
[3] https://github.com/scylladb/seastar/issues/2479#issuecomment-2427098413

No backport: not a regression.

### CODE REVIEW

**Comment by avikivity** [comment]
Missing: cmake. @tchaikov how to do that?

File: configure.py [hunk 1]
```diff
@@ -1662,6 +1662,9 @@ def dynamic_linker_option():
 curdir = os.getcwd()
 user_cflags = args.user_cflags + f" -ffile-prefix-map={curdir}=."
 
+# Since gcc 13, libgcc doesn't need the exception workaround
+user_cflags += ' -DNO_EXCEPTION_INTERCEPT -DSEASTAR_NO_EXCEPTION_HACK'
```
Comment (tchaikov) [inline configure.py:1666]: probably unrelated, but can we fix this on Seastar's end? because Seastar only supports the latest two major versions of compiler. when it comes to GCC, the latest two major versions are GCC-13 and GCC-14. both of them include the fix in libgcc of https://github.com/gcc-mirror/gcc/commit/6e80a1d164d1f996ad08a512c000025a7c2ca893#diff-03e90e365ad13835bc93337b2a1ccc264c121395d6d1a293c13948c8faf73d93R169 .

in general, user installs the runtime along with the compiler, presumably, we can assume that GCC-13 comes with libgcc-13. so i guess, we should can just either just drop the workaround, or define there two macros by default? if we really want to check the used runtime, probably we can use the proposed approach at https://github.com/scylladb/seastar/issues/2479#issuecomment-2428150691 ?

**Comment by tchaikov** [comment]
> Missing: cmake. @tchaikov how to do that?

tl;dr: it's already taken care of by your change.

long story: these two macros only take effect when building Seastar, and currently, we populate `user_cflags` to `modes[mode]['lib_cflags']`, which is in turn used for configuring Seastar's building system in `configure_seastar()`. this function is used by both `configure_using_cmake()` and `create_build_system()` after 415c83fa67e6b8434112c8e3f24830a3ff3d39f4 .

File: configure.py (hunk 1 above)
Comment (xemul) [reply configure.py:1666]: AFAICS with NO_EXCEPTION_INTERCEPT we'll also miss the `memory_cpp_exceptions` metrics accounting. Is that what we want?

Comment (tchaikov) [reply configure.py:1666]: ahh, i forgot that we also had non-multi-config-generator use case, so we need 
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 978b21686f..7b93b61a61 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -95,6 +95,10 @@ else()
     set(Seastar_SCHEDULING_GROUPS_COUNT 16 CACHE STRING "" FORCE)
     set(Seastar_UNUSED_RESULT_ERROR ON CACHE BOOL "" FORCE)
     add_subdirectory(seastar)
+    target_compile_definitions (seastar
+      PRIVATE
+        NO_EXCEPTION_INTERCEPT
+        SEASTAR_NO_EXCEPTION_HACK)
 endif()
 
 set(ABSL_PROPAGATE_CXX_STD ON CACHE BOOL "" FORCE)
```

Comment (avikivity) [reply configure.py]: > AFAICS with NO_EXCEPTION_INTERCEPT we'll also miss the `memory_cpp_exceptions` metrics accounting. Is that what we want?

No.

**Comment by avikivity** [comment]
v2: drop NO_EXCEPTION_INTERCEPT, it's unneeded and removes a useful feature

**Review by nyh — APPROVED** [review]

File: configure.py (hunk 1 above)
Comment (tchaikov) [reply configure.py]: created #21233
